Topic/2023.4/readme (#3)
metalama/Metalama.Compiler: README.md

<p align="center">
<img width="450" src="https://github.com/postsharp/Metalama/raw/master/images/metalama-by-postsharp.svg" alt="Metalama logo">
</p>

Metalama Compiler is a fork of [Roslyn](https://github.com/dotnet/Roslyn) that adds an extension points to allow analyzer projects to perform arbitrary transformations through the `ISourceTransformer` interface.

Metalama Compiler is actively and professionally maintained by [PostSharp Technologies](https://www.postsharp.net/).

Metalama Compiler is a part of [Metalama](https://github.com/postsharp/Metalama). Metalama Framework is a high-level meta-programming framework for C#. It's first pillar, boilerplate reduction, relies on Metalama Compiler to implement aspect-oriented programming. Metalama relies on several other Roslyn extension points: analyzers, diagnostic suppressors, source generators, code fix providers, and code refactoring providers. However, no extension point was available to transform source code during compilation, so we are relying on this fork.

> For an overview of the modification of Metalama Compiler compared to the vanilla Roslyn, see [this article](docs-Metalama\Modifications.md).

Reviewer: Backslash doesn't seem to work on GitHub.
```suggestion
For an overview of the modification of Metalama Compiler compared to the vanilla Roslyn, see [this article](docs-Metalama/Modifications.md).
```

## Features

Metalama Compiler adds the following features to Roslyn:

* **Source transformers.** Ability for analyzer projects to perform arbitrary transformations of source code during compilation thanks to the [ISourceTransformer](https://doc.metalama.net/api/metalama_compiler_isourcetransformer) interface.

* **Ordering.** Multiple source transformers can be ordered using the [TransformerOrderAttribute](https://doc.metalama.net/api/metalama_compiler_transformerorderattribute)  assembly-level custom attribute or the `MetalamaCompilerTransformerOrder` MSBuild property (see below)
    
    > [!Warning]
    > Having too many source transformers in a project can cause performance issues.

> * **Source mapping.** Diagnostics and PDBs are mapped to source code by default (even when reported on transformed code). This behavior can be changed by one of the new [MSBuild properties](docs-Metalama/Properties.md) defined by Metalama Compiler. This is by far the most complex feature of Metalama Compiler.

Reviewer: This link points to a file that you're deleting.

```suggestion
* **Source mapping.** Diagnostics and PDBs are mapped to source code by default (even when reported on transformed code). This behavior can be changed by one of the new [MSBuild properties](#msbuild-properties) defined by Metalama Compiler. This is by far the most complex feature of Metalama Compiler.
```

3. Add the `[Transformer]` custom attribute to this class.
4. Implement the `Execute` method. This method receives a `TransformerContext` object. Use this object to inspect the current compilation, add syntax trees, modify syntax trees, report diagnostics, or suppress diagnostics.

```cs
using Metalama.Compiler;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;

namespace Metalama.SourceTransformer
{
    [Transformer]
    internal class SourceTransformer : ISourceTransformer
    {
        public void Execute( TransformerContext context )
        {
           
        }
    }
}
```

### Step 2. Packaging a source transformer

The packaging of your transformer should follow the same rules as analyzers and source generators. That is, your output `dll` should be in the `analyzers/dotnet/cs` folder instead of under `lib`.

Follow [this example](https://github.com/dotnet/roslyn-sdk/blob/main/samples/CSharp/Analyzers/Analyzers.Implementation/Analyzers.CSharp.csproj) from the Roslyn documentation:

1. Define the `IncludeBuildOutput=False` property to prevent the project output to be added to the `lib` directory of the package.

> 2. Use the following to include the output under the `analysis/dotnet/cs`:

Reviewer: ```suggestion
2. Use the following to include the output under the `analyzers/dotnet/cs` path:
```
